1.MLP_Binary_Classifier/MLP_utils/utils.py

```diff
@@ -0,0 +1,627 @@
+    optimizer = getattr(optim, optimization_params["optimizer"])(
+        model.parameters(), lr=optimization_params["learning_rate"]
+    )
+    # loss function
+
+    # for binary model use different for multi-class
+    criterion = nn.BCEWithLogitsLoss()
+
+    # send model to device(cuda)
+
+    model = model.to(params.DEVICE)
+    criterion = criterion.to(params.DEVICE)
+
+    # train set accuracy and loss
+    train_acc = []
+    train_loss = []
+
+    # validation set accuracy and loss
+    valid_acc = []
+    valid_loss = []
+
+    # total number of data to pass through
+    total_step = len(train_loader)
+    total_step_val = len(valid_loader)
+
+    for epoch in range(params.OPTIM_EPOCHS):
+
+        (
+            train_acc,
+            train_loss,
+            valid_acc,
+            valid_loss,
+            correct,
+            total,
+            correct_v,
+            total_v,
+        ) = train_n_validate(
+            model,
+            optimizer,
+            criterion,
+            train_acc,
+            train_loss,
+            valid_acc,
+            valid_loss,
+            total_step,
+            total_step_val,
+            params,
+            train_loader,
+            valid_loader,
+        )
+
+        trial.report(np.mean(valid_loss), epoch)
+
+        # Handle pruning based on the intermediate value
+        if trial.should_prune():
+            raise optuna.exceptions.TrialPruned()
+
+    # I want information returned but only 1 metric required for the optimize function called by study.optimize
+    # with out this conditional statement the optimization will fail
+    if return_info == True:
+        print(f"Validation Accuracy: {np.mean(valid_acc)}")
+        print(f"Validation Loss: {np.mean(valid_loss)}")
+        print(f"Training Accuracy: {np.mean(train_acc)}")
+        print(f"Training Loss: {np.mean(train_loss)}")
+    else:
+        return np.mean(valid_loss)
+
+
+def extract_best_trial_params(best_params):
+    """
+    This function extractions the best parameters from the best trial.
+
+    These extracted parameters will be used to create a new model.
+
+    Parameters:
+        best_params : obj
+            study.best_params
+            the best_params function of the study object
+    Return:
+        param_dict : dict
+            dictionary of all of the params for the best trial model
+    """
+
+    params = best_params
+    units = []
+    dropout = []
+    n_layers = params["n_layers"]
+    optimizer = params["optimizer"]
+    lr = params["learning_rate"]
+    for i in range(params["n_layers"]):
+        units.append(params[f"n_units_l{i}"])
+        dropout.append(params[f"dropout_{i}"])
+    param_dict = {
+        "units": units,
+        "dropout": dropout,
+        "n_layers": n_layers,
+        "optimizer": optimizer,
+        "learning_rate": lr,
+    }
+    return param_dict
+
+
+# function for new optimized model
+def optimized_model(in_features, paramater_dict):
```

Comment:
I'm confused about what this function does - it doesn't seem to optimize a model, but maybe I'm reading it wrong? Is there a better name to reduce confusion?